feat(http): exclude caching for authenticated HTTP requests

alan-agius4 opened the pull request:
This update modifies the transfer cache logic to prevent caching of HTTP requests that require authorization. To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`

BREAKING CHANGE: By default we now prevent caching of HTTP requests that require authorization . To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`.

Example:
```ts
withHttpTransferCache({
  includeRequestsWithAuthHeaders: true,
})
```

Closes: #54745

AndrewKushnir reviewed (commented):
@alan-agius4 thanks for the PR 👍 I've added a comment about the name of the option.

We can also add a note about this in the SSR guide, WDYT? Here is a proposed change:

```
--- a/adev/src/content/guide/ssr.md
+++ b/adev/src/content/guide/ssr.md
@@ -71,7 +71,7 @@ Hydration is the process that restores the server side rendered application on t
 
 [`HttpClient`](api/common/http/HttpClient) cached outgoing network requests when running on the server. This information is serialized and transferred to the browser as part of the initial HTML sent from the server. In the browser, `HttpClient` checks whether it has data in the cache and if so, reuses it instead of making a new HTTP request during initial application rendering. `HttpClient` stops using the cache once an application becomes [stable](api/core/ApplicationRef#isStable) while running in a browser.
 
-`HttpClient` caches all `HEAD` and `GET` requests by default. You can configure the cache by using [`withHttpTransferCacheOptions`](/api/platform-browser/withHttpTransferCacheOptions) when providing hydration.
+By default, `HttpClient` caches all `HEAD` and `GET` requests which don't contain `Authorization` or `Proxy-Authorization` headers. You can override those settings by using [`withHttpTransferCacheOptions`](/api/platform-browser/withHttpTransferCacheOptions) when providing hydration.
 
 <docs-code language="typescript">
```

JoostK commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -97,13 +100,13 @@ export function transferCacheInterceptorFn(
+    // Do not cache request that require authorization when includeAuthorizationRequests is falsey
+    (!globalOptions.includeAuthorizationRequests &&
+      (req.headers.has('authorization') || req.headers.has('proxy-authorization'))) ||
     globalOptions.filter?.(req) === false
```
Given that there's already global filtering capabilities available, do we really want to handle Authorization as special case?

alan-agius4 replied:
The concept was to prioritize doing the "right" thing by default, which implies that `authorization` requests won't be cached by default.

Considering this, perhaps the global filter capabilities should execute before any other checks. If it returns `true`, subsequent internal checks could be bypassed.

Consequently, this approach could render the `includePostRequest` option unnecessary.

//cc @AndrewKushnir

alan-agius4 committed:
feat(http): exclude caching for authenticated HTTP requests

This update modifies the transfer cache logic to prevent caching of HTTP requests that require authorization. To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`

BREAKING CHANGE: By default we now prevent caching of HTTP requests that require authorization . To opt-out from this behaviour use the `includeRequestsWithAuthHeaders` option in `withHttpTransferCache`.

Example:
```ts
withHttpTransferCache({
  includeRequestsWithAuthHeaders: true,
})
```

Closes: #54745

AndrewKushnir commented on packages/common/http/src/transfer_cache.ts:
```diff
@@ -100,10 +103,13 @@ export function transferCacheInterceptorFn(
     (requestMethod !== 'POST' && !ALLOWED_METHODS.includes(requestMethod)) ||
-    requestOptions === false || //
+    // Do not cache request that require authorization when includeRequestsWithAuthHeaders is falsey
+    (!globalOptions.includeRequestsWithAuthHeaders &&
+      (req.headers.has('authorization') || req.headers.has('proxy-authorization'))) ||
```
nit: I'd propose moving the logic that checks for specific headers into a separate function:
```suggestion
    (!globalOptions.includeRequestsWithAuthHeaders && hasAuthHeaders(req)) ||
```

AndrewKushnir reviewed (approved):
Looks great, just 1 minor comment 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

alan-agius4 committed:
fixup! feat(http): exclude caching for authenticated HTTP requests

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

alxhub commented:
This PR was merged into the repository by commit 8eacb6e4b982a5aa23cfbf9078dc4e19d9466d73.
